### PR TITLE
Added tests for detecting references between summaries

### PR DESCRIPTION
**Pull request by agarwal-navin**
Added tests for validating that we can detect object references between summaries and update the unreferenced timestamp for such objects (if applicable).
Currently, we cannot detect these references and so the tests don't work as expected. I added a function `assertTestFails` that asserts that the final test result infact fails. This is a setup for the this GC work item - https://github.com/microsoft/FluidFramework/issues/7924.

Basically, these tests fail now and once the above work item is completed, they should pass (the condition in `assertTestFails` will be reverted).

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts [hunk 1]
```diff
@@ -132,165 +132,461 @@ describeFullCompat("GC unreferenced timestamp", (getTestObjectProvider) => {
...
+     * The nodes are data stores represented by alphabets A, B, C and so on.
+     */
+    describe("References between summaries - state transition from unreferenced -> referenced -> unreferenced", () => {
+        /**
+         * Function that asserts the given test result fails. This is because all the scenarios here currently fail.
+         * These should pass once this issue is fixed - https://github.com/microsoft/FluidFramework/issues/7924. The
+         * assert condition will be flipped then.
+         */
+         function assertTestFails(testResult: boolean, message: string) {
+            assert(!testResult, message);
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:215]: Note this function reverts the test result which basically means that the tests don't work as expected now.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 2]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
+        const nodeC = "/C";
+        const nodeD = "/D";
+
+        /**
+         * Function that asserts the given test result fails. This is because all the scenarios here currently fail.
+         * These should pass once this issue is fixed - https://github.com/microsoft/FluidFramework/issues/7924. The
+         * assert condition will be flipped then.
+         */
+        function assertTestFails(testResult: boolean, message: string) {
+            assert(!testResult, message);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:391]: Note this function reverts the test result which basically means that the tests don't work as expected now.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 3]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
-    beforeEach(async () => {
-        mockLogger = new MockLogger();
+    describe("Inactive events", () => {
+        const inactiveObjectRevivedEvent = "GarbageCollector:inactiveObjectRevived";
+        const inactiveObjectChangedEvent = "GarbageCollector:inactiveObjectChanged";
+
+        // Waits for > deleteTimeoutMs. To be called to make sure that any unreferenced nodes have been deleted.
+        async function waitForDeleteTimeout(): Promise<void> {
+            await new Promise<void>((resolve) => {
+                setTimeout(resolve, deleteTimeoutMs + 100);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:78]: It would be great to mock timeouts and make sure test code does not need to actually sit idle.
I think we have an example of that somewhere in the code. Search for sinon.useFakeTimers()

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 4]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
+        function changeAllNodes(garbageCollector: IGarbageCollector) {
+            nodes.forEach((nodeId) => {
+                garbageCollector.nodeChanged(nodeId);
+            });
+        }
+
+        // Returns a dummy snapshot tree to be built upon.
+        const getDummySnapshotTree = (): ISnapshotTree => {
+            return {
+                id: "dummy",
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:103]: It would be great not to have IDs as I think we should delete it from ISnapshotTree (as there is no usage for it).
So, I marked it as optional some time ago but did not get to deleting it. Would be great for our code not to add IDs.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 5]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
+
+            // Wait for unreferenced timer (if any) to expire.
+            await waitForDeleteTimeout();
+
+            // Change all nodes. This should result in an inactiveObjectChanged event for node 2 and node 3 since they
+            // are inactive.
+            changeAllNodes(garbageCollector);
+            assert(
+                mockLogger.matchEvents([
+                    { eventName: inactiveObjectChangedEvent, deleteTimeoutMs, inactiveNodeId: nodes[2] },
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:163]: I'd prefer for us to use smaller set of fields/properties in telemetry, as it otherwise explodes, and affects storage cost, but also ease of authoring simple queries. Can we use 'duration' ?

Comment (vladsud) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:163]: Somewhat similar for inactiveNodeId - can it be `id`?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 6]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
-        validateNoInactiveEvents();
+            await waitForDeleteTimeout();
 
-        // Wait for unreferenced timer (if any) to expire.
-        await waitForDeleteTimeout();
+            // Change all nodes. This should result in an inactiveObjectChanged event for node 2 and node 3 since they
+            // are inactive.
+            changeAllNodes(garbageCollector);
+            assert(
+                mockLogger.matchEvents([
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:196]: Something to consider logging structure can change over time. There are good and bad side-effects of using logging for testing. In some cases, it's better to have direct API / eventing system. It might be _public_ in a sense of particular layer (GC), but not exposed through other layers (like container runtime), thus keeping it mostly private / for internal FF usage.
I do not know if that's that case or not, just speaking out loudly :)

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 7]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
 
-        // Add reference to node 3 from node 1.
-        defaultGCData.gcNodes[nodes[1]] = [ nodes[3] ];
+        it("generates inactive events for nodes that are inactive on load", async () => {
+            // Create GC state where node 3's unreferenced time was > deleteTimeoutMs ago.
+            // This means this node should become inactive as soon as its data is loaded.
+
+            // Create a snapshot tree to be used as the GC snapshot tree.
+            const gcSnapshotTree = getDummySnapshotTree();
+            const gcBlobId = "gc_blob";
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:214]: Should this value be in sync with some value in code? Or it can be any value? I can't find it in code otherwise.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 8]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
+            await garbageCollector.collectGarbage({ runGC: true });
+            assert(
+                mockLogger.matchEvents([
+                    { eventName: inactiveObjectRevivedEvent, deleteTimeoutMs, inactiveNodeId: nodes[3] },
+                ]),
+                "inactiveObjectRevived event not generated as expected",
+            );
+        });
+
+        it(`generates inactive events for nodes that are inactive on load - GC state is present in multiple` +
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:313]: Consider making names shorter, and instead putting more detailed description in comment above it.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:391]: Please add issue # in the code that tracks it (even if it's next PR :))

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 9]
```diff
@@ -104,266 +68,583 @@ describe("Garbage Collection Tests", () => {
...
+         * Function that asserts the given test result fails. This is because all the scenarios here currently fail.
+         * These should pass once this issue is fixed - https://github.com/microsoft/FluidFramework/issues/7924. The
+         * assert condition will be flipped then.
+         */
+        function assertTestFails(testResult: boolean, message: string) {
+            assert(!testResult, message);
+        }
+
+        // Adds a small delay between two GC runs so that the unreferenced timestamp can be updated.
+        async function addDelay(): Promise<void> {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:395]: We already have delay() function in common-utils, so this one just adds value to arg, but no need to create your own promises / setTimeout.

File: packages/test/test-end-to-end-tests/src/packageVersion.ts [hunk 10]
```diff
@@ -6,4 +6,4 @@
  */
 
 export const pkgName = "@fluidframework/test-end-to-end-tests";
-export const pkgVersion = "0.53.0";
+export const pkgVersion = "0.54.0";
```
Comment (vladsud) [inline packages/test/test-end-to-end-tests/src/packageVersion.ts:9]: You should not be bumping this in your PRs. Is this a manifest of improperly done bump?

File: packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts [hunk 11]
```diff
@@ -132,165 +132,461 @@ describeFullCompat("GC unreferenced timestamp", (getTestObjectProvider) => {
...
+            assert(dsBTimestamp3 !== undefined, `data store should still have unreferenced timestamp`);
+            assert.strictEqual(dsBTimestamp2, dsBTimestamp3, "unreferenced timestamp should not have changed");
+
+            // Mark the data store as referenced again and validate that the unreferenced timestamp is removed.
+            dataStoreA._root.set("dataStoreB", dataStoreB.handle);
+            // Validate that the data store does not have unreferenced timestamp after being referenced.
+            const timestamps4 = await getUnreferencedTimestamps(summarizerClient);
+            const dsBTimestamp4 = timestamps4.get(dataStoreB.id);
+            assert(dsBTimestamp4 === undefined, `data store should not have unreferenced timestamp anymore`);
+        }).timeout(20000);
```
Comment (vladsud) [inline packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:169]: Please, no 20 second running tests! Please mock timeouts and figure out how to run it in 2 seconds.

Comment (vladsud) [reply packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:169]: Same below

File: packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts [hunk 12]
```diff
@@ -132,165 +132,461 @@ describeFullCompat("GC unreferenced timestamp", (getTestObjectProvider) => {
...
+
+        /**
+         * Validates that references added by unreferences nodes do not show up as references.
+         * 1. Summary 1 at t1. V = [A*, B, C]. E = []. B and C have unreferenced time t1.
+         * 2. Op adds reference from B to C. E = [B -> C].
+         * 3. Summary 2 at t2. V = [A*, B, C]. E = [B -> C]. B and C have unreferenced time t1.
+         * Validates that the unreferenced time for B and C is still t1.
+         */
+         it(`Scenario 8 - An unreferenced node B adds reference to another node C`, async () => {
```
Comment (vladsud) [inline packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:559]: Is there a chance to reuse some code between these scenarios?

**Review by vladsud — APPROVED** [review]
I did not look very closely at substance of the tests, only at structure and general usage.
Please figure out how not to do delays in tests.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:391]: There is issue# in the comment of this function :-).

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:214]: This is the the blob id that is generated when a blob is uploaded. The snapshot tree that is downloaded has a mapping of "blob key" to "blob id". The key here is anything that starts with "gcBlobPrefix" which matches the code.
Added a comment to clarify some of it.

File: packages/test/test-end-to-end-tests/src/packageVersion.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/packageVersion.ts:9]: I think this happened because of merge. My change was in flight and we released.

File: packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts (hunk 11 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts:169]: This is not due to delays added in tests. The tests take long to run because it uploads summary to odsp and then waits for the ack. Each test does this multiple times so it takes pretty long. Is there a good way to reduce this time?

File: packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts]: I don't think so. I moved out the common parts already. Maybe I can squeeze some more things together but it would make the tests harder to read.

**Comment by agarwal-navin** [comment]
> I did not look very closely at substance of the tests, only at structure and general usage. Please figure out how not to do delays in tests.

I fixed the unit tests to use fake timers. For the end-to-end tests, there are a few tests that use delays, I created an issue to fix them - https://github.com/microsoft/FluidFramework/issues/8553

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts]: Removed delay and added mock timers.
